marlowe/test/Spec/Marlowe/Semantics/Arbitrary.hs

```diff
@@ -0,0 +1,993 @@
+                      -> Context                           -- ^ The Marlowe context.
+                      -> Gen (Case Contract)               -- ^ Action for generating the case.
+arbitraryCaseWeighted w context =
+  Case <$> semiArbitrary context <*> arbitraryContractWeighted w context
+
+
+-- | Generate one of the golden contracts and its initial state.
+goldenContract :: Gen (Contract, State)
+goldenContract = (,) <$> elements goldenContracts <*> pure (State AM.empty AM.empty AM.empty $ POSIXTime 0)
+
+
+instance Arbitrary Contract where
+  arbitrary = frequency [(95, semiArbitrary =<< arbitrary), (5, fst <$> goldenContract)]
+  shrink (Pay a p t x c) = [Pay a' p t x c | a' <- shrink a] ++ [Pay a p' t x c | p' <- shrink p] ++ [Pay a p t' x c | t' <- shrink t] ++ [Pay a p t x' c | x' <- shrink x] ++ [Pay a p t x c' | c' <- shrink c]
+  shrink (If o x y) = [If o' x y | o' <- shrink o] ++ [If o x' y | x' <- shrink x] ++ [If o x y' | y' <- shrink y]
+  shrink (When a t c) = [When a' t c | a' <- shrink a] ++ [When a t' c | t' <- shrink t] ++ [When a t c' | c' <- shrink c]
+  shrink (Let v x c) = [Let v' x c | v' <- shrink v] ++ [Let v x' c | x' <- shrink x] ++ [Let v x c' | c' <- shrink c]
+  shrink (Assert o c) = [Assert o' c | o' <- shrink o] ++ [Assert o c' | c' <- shrink c]
+  shrink _ = []
+
+
+-- | Generate an arbitrary contract, weighted towards different contract types.
+arbitraryContractWeighted :: [(Int, Int, Int, Int, Int, Int)]  -- ^ The weights of contract terms.
+                          -> Context                           -- ^ The Marlowe context.
+                          -> Gen Contract                      -- ^ Action for generating the contract.
+arbitraryContractWeighted ((wClose, wPay, wIf, wWhen, wLet, wAssert) : w) context =
+  frequency
+    [
+      (wClose , pure Close)
+    , (wPay   , Pay <$> semiArbitrary context <*> semiArbitrary context <*> semiArbitrary context <*> semiArbitrary context <*> arbitraryContractWeighted w context)
+    , (wIf    , If <$> semiArbitrary context <*> arbitraryContractWeighted w context <*> arbitraryContractWeighted w context)
+    , (wWhen  , When <$> listOf (arbitraryCaseWeighted w context) `suchThat` ((<= length w) . length) <*> semiArbitrary context <*> arbitraryContractWeighted w context)
+    , (wLet   , Let <$> semiArbitrary context <*> semiArbitrary context <*> arbitraryContractWeighted w context)
+    , (wAssert, Assert <$> semiArbitrary context <*> arbitraryContractWeighted w context)
+    ]
+arbitraryContractWeighted [] _ = pure Close
+
+-- | Default weights for contract terms.
+defaultContractWeights :: (Int, Int, Int, Int, Int, Int)
+defaultContractWeights = (35, 20, 10, 15, 20, 5)
+
+-- | Contract weights selecting only `Close`.
+closeContractWeights :: (Int, Int, Int, Int, Int, Int)
+closeContractWeights = (1, 0, 0, 0, 0, 0)
+
+-- | Contract weights selecting only `Pay`.
+payContractWeights :: (Int, Int, Int, Int, Int, Int)
+payContractWeights = (0, 1, 0, 0, 0, 0)
```

Comment:
They have to be used carefully, but they are a provided as a list. For example, it is safe to use `payContractWeights : repeat defaultContractWeights` because that will contain a `Close`. It would be unsafe to use `repeat payContractWeights` because that doesn't have a `Close`. Some of the test cases require specific constructs for the first two levels of the contract and then the continuations after that are arbitrary. I'll add a comment to warn user that they need to have some weight for close eventually.

Reply:
Ah, I see, that is convenient

Reply:
Fixed: https://github.com/input-output-hk/marlowe-cardano/blob/8e9144b8c75eecf8a9d7dd3e437bea6cbc686f65/marlowe/test/Spec/Marlowe/Semantics/Arbitrary.hs#L755